nav2_bt_navigator/src/bt_navigator.cpp

```diff
@@ -229,10 +230,49 @@ BtNavigator::onLoop()
 }
 
 void
-BtNavigator::onPreempt()
+BtNavigator::onPreempt(Action::Goal::ConstSharedPtr goal)
 {
   RCLCPP_INFO(get_logger(), "Received goal preemption request");
-  initializeGoalPose(bt_action_server_->acceptPendingGoal());
+
+  if (goal->behavior_tree == bt_action_server_->getCurrentBTFilename() ||
+    (goal->behavior_tree.empty() &&
+    bt_action_server_->getCurrentBTFilename() == bt_action_server_->getDefaultBTFilename()))
+  {
+    // if pending goal requests the same BT as the current goal, accept the pending goal
+    // if pending goal has an empty behavior_tree field, it requests the default BT file
+    // accept the pending goal if the current goal is running the default BT file
+    initializeGoalPose(bt_action_server_->acceptPendingGoal());
+  }
+}
+
+bool
+BtNavigator::shouldCancelCurrentGoal()
+{
+  if (!bt_action_server_->isPreemptRequested()) {
+    return false;
+  }
+
+  auto pending = bt_action_server_->getPendingGoal();
+
+  if (pending) {
+    // if pending goal requests the same BT as the current goal, don't cancel the current goal
+    // the old BT will be used to handle the pending goal
+    if (pending->behavior_tree == bt_action_server_->getCurrentBTFilename() ||
+      (pending->behavior_tree.empty() &&
+      bt_action_server_->getCurrentBTFilename() == bt_action_server_->getDefaultBTFilename()))
+    {
+      return false;
+    }
+
+    // if pending goal requested a different BT we'll cancel the current goal
+    RCLCPP_INFO(
+      get_logger(),
+      "Goal was preempted with a different BT XML file, cancelling the previous goal.");
+    return true;
```

Comment:
This all logically tracks for me, but really makes me wonder if this is OK. What quite literally we're doing is saying:
- Do you want to preempt with a new BT?
- OK, we're going to cancel and reissue 

Wouldn't we, in this situation, rather the user cancel themselves and issue a new goal when changing BTs? This isn't really "preemption" if we're actually cancelling, its not a continuous, smooth transition from BT 1 to BT 2. 

It makes me think that if this is the only clear way to enable preemption (by canceling and accepting the pending goal), we shouldn't support preemption of new BTs at all (e.g. the `onPreempt` if not the same BT should reject the goal.